### PR TITLE
[browser][MT] invocation of user JS in UI thread till next event loop tick

### PR DESCRIPTION
**Pull request by pavelsavara**
postpone invocation of user JS in UI thread till next event loop tick, so that `em_task_queue_execute` is not blocked by it. 
That allows `emscripten_yield` to process FS POSIX calls even if the UI is spin-waiting for the return from synchonous call to `JSExport`

 - new args field `SyncDoneSemaphorePtr`
 - new `invoke_later_when_on_ui_thread_sync`
 - new `invoke_later_when_on_ui_thread_async` which postpone invocation of user code till next event loop tick
 - rename `mono_wasm_invoke_jsimport_MT`
 - test in `browser-threads` sample, more tests in next PR

### CODE REVIEW

File: src/mono/browser/runtime/driver.c [hunk 1]
```diff
@@ -256,7 +256,7 @@ mono_wasm_invoke_jsexport (MonoMethod *method, void* args)
 #ifndef DISABLE_THREADS
 
 extern void mono_threads_wasm_async_run_in_target_thread_vii (void* target_thread, void (*func) (gpointer, gpointer), gpointer user_data1, gpointer user_data2);
-extern void mono_threads_wasm_sync_run_in_target_thread_vii (void* target_thread, void (*func) (gpointer, gpointer), gpointer user_data1, gpointer user_data2);
+extern void mono_threads_wasm_sync_run_in_target_thread_vii (void* target_thread, void (*func) (gpointer, gpointer), gpointer user_data1, gpointer args);
```
Comment (radekdoulik) [inline src/mono/browser/runtime/driver.c:259]: nit: shouldn't the last parameter remain named `user_data2` as in the function above?

**Review by radekdoulik — APPROVED** [review]

File: src/mono/browser/runtime/marshal-to-js.ts [hunk 2]
```diff
@@ -338,6 +339,10 @@ function create_task_holder(res_converter?: MarshalerToJs) {
 }
 
 export function mono_wasm_resolve_or_reject_promise(args: JSMarshalerArguments): void {
+    // rejection/resolution should not arrive earlier than the promise created by marshaling in mono_wasm_invoke_jsimport_MT 
+    invoke_later_when_on_ui_thread_async(() => mono_wasm_resolve_or_reject_promise_impl(args));
+}
+export function mono_wasm_resolve_or_reject_promise_impl(args: JSMarshalerArguments): void {
```
Comment (maraf) [inline src/mono/browser/runtime/marshal-to-js.ts:345]: nit: Same name pattern as above?

```suggestion
export function mono_wasm_resolve_or_reject_promise_ui(args: JSMarshalerArguments): void {
```

File: src/mono/sample/wasm/browser-threads/Program.cs
Comment (maraf) [inline src/mono/sample/wasm/browser-threads/Program.cs:1]: Would be great to rewrite the sample as WBT test app 🤞

**Review by maraf — APPROVED** [review]

File: src/mono/browser/runtime/marshal-to-js.ts (hunk 2 above)
Comment (pavelsavara) [reply src/mono/browser/runtime/marshal-to-js.ts:345]: mono_wasm_resolve_or_reject_promise_impl is also called on WebWorker ... so no rename

**Comment by pavelsavara** [comment]
/ba-g there is [Log](https://helix.dot.net/api/2019-06-17/jobs/950f5108-0f44-422b-b0b7-7f307ae59e85/workitems/Workloads-NoWebcil-Wasm.Build.Tests.TestAppScenarios.SignalRClientTests/console) which is known issue https://github.com/dotnet/runtime/issues/99961

**Pull request merged**